pynhm/base/budget.py

```diff
@@ -0,0 +1,172 @@
+from warnings import warn
+
+import numpy as np
+
+from ..constants import zero
+from .accessor import Accessor
+
+# # For both 1) cumulative (start_date), and 2) current time step
+# # * Individual input terms: name, units, & subclass
+# # * Individual output terms: name, units, & subclass
+# # * Sum input terms
+# # * Sum output terms
+# # In - out
+# # ? check against a reference storage?
+
+# budget needs to be able to restart?
+# should storage unit edit its own metadata to set it's subclass on each variable?
+
+# the input dictionarys are a strange stucture
+# key: {data: np.ndarray([nhru]), **metadata}
+# might consider other ways of passing the data.
+# ** should carry the data and the meta separately ** since that's what we do elsewhere
+
+# apply initial accumulations passed via e.g.
+# initial_accumulations = {
+#     "inputs": {"var": data, ...},
+#     "outputs": {"var": data, ...},
+#     "storage_changes": {"var": data, ...}, }
+
+
+class Budget(Accessor):
```

Comment:
Would be nice to have some sort of output method that prints a modflow-like budget table to a text file.  In modflow, we also have csv output that writes the inputs, outputs, storage change, percent discrepancy, etc. to a csv.  This can wait, though.

Reply:
yes. i was going to define a __repr__ i was also thinking writing to a budget "group" in a netcdf file

Reply:
Repr is in place. Here is an example: 

```
*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-
Budget (units: None) of PRMSCanopy for entire model domain (spatial sum)
@ time: 1979-05-31T00:00:00 (itime_step: 150)

This timestep, rates:
         input rates             output rates                 storage change rates            
         --------------------    -------------------------    --------------------------------
         hru_rain: 72.6598...    net_rain: 72.659867992...    hru_intcpstor_change: -9.5858...        
         hru_snow: 0.0           net_snow: 0.0                                                        
                                 hru_intcpevap: 9.58585...                                            
         -------------------------------------------------------------------------------------
Balance:           72.6598...    -              82.2457...    =                     -9.5858...    

Accumulated volumes (since 1978-12-31T00:00:00):
         input volumes           output volumes               storage change volumes   
         --------------------    -------------------------    --------------------------------
         hru_rain: 16464.0...    net_rain: 16228.819585...    hru_intcpstor_change: 3.04456...        
         hru_snow: 2598.06...    net_snow: 2605.6412818...                                            
                                 hru_intcpevap: 224.585...                                            
         -------------------------------------------------------------------------------------
Balance:           19062.0...    -              19059.0...    =                     3.04456...    

```

Not perfect but a great start